### PR TITLE
one-to-many criteria to goals relation

### LINKED ISSUE #1752: Project score is incorrect
Currently we show outdated score

### LINKED ISSUE #2000: one-to-many criteria to goals relation

### PR DESCRIPTION
**Pull request by LamaEats**
Now can be link one goal to many criterias
Added a new column `criteriaGoalId` to `GoalAchieveCriteria` model
Added a new relation `goalInCriteria` between `Goal` and `GoalAchieveCriteria` model

## PR includes

- [x] Bug Fix

## Related issues

Resolve #2000
Resolve #1752

### CODE REVIEW

File: src/components/GoalCriteria/GoalCriteria.tsx [hunk 1]
```diff
@@ -170,12 +170,12 @@ const GoalCriteriaItem: React.FC<GoalCriteriaItemProps> = (props) => {
...
-    const itemToRender: Partial<GoalAchiveCriteria['goalAsCriteria'] & { shortId: string; weight: number }> =
+    const itemToRender: Partial<GoalAchiveCriteria['criteriaGoal'] & { shortId: string; weight: number }> =
         useMemo(() => {
-            if (item.goalAsCriteria) {
+            if (item.criteriaGoal) {
                 return {
-                    ...item.goalAsCriteria,
-                    shortId: `${item.goalAsCriteria.projectId}-${item.goalAsCriteria.scopeId}`,
+                    ...item.criteriaGoal,
+                    shortId: `${item.criteriaGoal.projectId}-${item.criteriaGoal.scopeId}`,
```
Comment (awinogradov) [inline src/components/GoalCriteria/GoalCriteria.tsx:178]: Must be on server-side

File: src/components/HistoryRecord/HistoryRecord.tsx [hunk 2]
```diff
@@ -455,20 +455,20 @@ export const HistoryRecordParticipant: React.FC<
...
         return (
             <StyledIssueListItem
                 strike={strike}
                 size="xs"
                 issue={{
-                    title: goalAsCriteria.title,
-                    _shortId: `${goalAsCriteria.projectId}-${goalAsCriteria.scopeId}`,
-                    id: goalAsCriteria.id,
+                    title: criteriaGoal.title,
+                    _shortId: `${criteriaGoal.projectId}-${criteriaGoal.scopeId}`,
```
Comment (awinogradov) [inline src/components/HistoryRecord/HistoryRecord.tsx:470]: Must be on server-side

File: src/utils/db.ts [hunk 3]
```diff
@@ -334,52 +334,82 @@ export const makeGoalRelationMap = <T extends Goal>(
...
                 },
             },
         },
     });
 
+    type CriteriaItem = (typeof criteriaList)[number];
+
+    const projectsToUpdate = new Set<string>();
+
+    const groupedCriteriaListByGoals = criteriaList.reduce<{ [key: string]: CriteriaItem[] }>((acc, criteria) => {
```
Comment (awinogradov) [inline src/utils/db.ts:362]: `Record<string, CriteriaItem[]>`

File: trpc/queries/goals.ts [hunk 4]
```diff
@@ -553,8 +553,11 @@ export const goalDeepQuery = {
 const maxPossibleCriteriaWeight = 100;
 
 export const calcAchievedWeight = (
-    list: (GoalAchieveCriteria & { goalAsCriteria: (Goal & { state: State | null }) | null })[],
+    list: (GoalAchieveCriteria & {
+        criteriaGoal: (Goal & { state: State | null }) | null;
+    })[],
 ): number => {
+    console.log(list);
```
Comment (awinogradov) [inline trpc/queries/goals.ts:560]: ;)

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/components/GoalCriteria/GoalCriteria.tsx (hunk 1 above)
Comment (LamaEats) [reply src/components/GoalCriteria/GoalCriteria.tsx:178]: Will fix in #1997

File: src/components/HistoryRecord/HistoryRecord.tsx (hunk 2 above)
Comment (LamaEats) [reply src/components/HistoryRecord/HistoryRecord.tsx:470]: will fix in in #2043

**Review by tsumo — APPROVED** [review]

File: trpc/queries/goals.ts [hunk 5]
```diff
@@ -598,8 +563,22 @@ export const addCalculatedGoalsFields = (goal: any, activityId: string, role: Ro
...
-            ? calcAchievedWeight(goal.goalAchieveCriteria)
+        goal.completedCriteriaWeight == null && goal.goalAchieveCriteria?.length
+            ? calcAchievedWeight(
+                  goal.goalAchieveCriteria.map(
+                      ({
+                          weight,
+                          isDone,
+                          deleted,
+                          criteriaGoal,
+                      }: GoalAchieveCriteria & { criteriaGoal: (Goal & { state: State | null }) | null }) => ({
```
Comment (asabotovich) [inline trpc/queries/goals.ts:574]: Can we use the same structure in calcAchievedWeight and avoid map for every usage?

**Pull request closed without merging**